patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs

```diff
@@ -0,0 +1,184 @@
+		/// <param name="mod">The mod who owns the recipe.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe = new ModRecipe(mod);
+			_recipe.SetResult(mod, itemName, stack);
+		}
+
+		/// <summary>Creates a new instance with no ingredients with the given item and stack as a result.</summary>
+		/// <param name="item"></param>
+		/// <param name="stack"></param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(ModItem item, int stack = 1)
+		{
+			_recipe = new ModRecipe(item.mod);
+			_recipe.SetResult(item, stack);
+		}
+
+
+		/// <summary>
+		/// Adds an ingredient to this recipe with the given item type and stack size.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe)</example>
+		/// </summary>
+		/// <param name="type">The item type.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(int type, int stack = 1)
+		{
+			_recipe.AddIngredient(type, stack);
+
+			return this;
+		}
+
+
+		// System.TupleValue is not resolved. Need to import it from a NuGet package ?
+		/*public RecipeBuilder Requires(params (short itemId, int stack)[] ingredients)
+		{
+			for (int i = 0; i < ingredients.Length; i++)
+				Requires(ingredients[i].Item1, ingredients[i].Item2);
+
+			return this;
+		}*/
+
+
+		/// <summary>
+		/// Adds the specified ingredients to this recipe with the given item types.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe)</example>
+		/// </summary>
+		/// <param name="type1">The first item type.</param>
+		/// <param name="type2">The second item type.</param>
+		/// <param name="type3">The third item type.</param>
+		/// <param name="types">The remaining item types.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(int type1, int type2, int type3, params int[] types)
+		{
+			Requires(type1);
+			Requires(type2);
+			Requires(type3);
+
+
+			for (int i = 0; i < types.Length; i++)
+				Requires(types[i]);
+
+
+			return this;
+		}
+
+
+		/// <summary>Adds an ingredient to this recipe with the given item name from the given mod, and with the given stack stack. If the mod parameter is null, then it will automatically use an item from the mod creating this recipe.</summary>
+		/// <param name="mod">The mod.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">The item " + itemName + " does not exist in mod " + mod.Name + ". If you are trying to use a vanilla item, try removing the first argument.</exception>
+		public RecipeBuilder Requires(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe.AddIngredient(mod ?? _recipe.mod, itemName, stack);
```

Comment:
Why the null coalescing? The player should pass mod, if it is incorrect there should be an error

Reply:
As per the TML documentation, if the passed mod is null, it'll use the one who created the recipe.

See http://tmodloader.github.io/tModLoader/html/class_terraria_1_1_mod_loader_1_1_mod_recipe.html#a7c390feb8cf0a3212de31b53ad5bcfc2.

I could let `AddIngredient` do the work if thats what it already does I suppose.

Reply:
Strings should only be usable for accessing other mods' content in such a manual context. Why is there even a string overload if there isn't one with a generic ModItem parameter?

Reply:
I'm not entirely sure what you mean. I followed the existing structure that ModRecipe provides. Do you mean to say we should have a `Require<T>(...) where T : ModItem` ?